### PR TITLE
feat: WPS Implementation for UAE and Qatar and some generic fields for Employee master.

### PR DESCRIPTION
**Pull request by Anurag810**
### WPS Implementaion for UAE and Qatar
The Wages Protection System (“WPS”) has operated for several years and served as a useful means of codifying, regulating, and ensuring the timely payment of employee salaries as well as providing an effective method for monitoring employer non-compliance. Where Hr user sends Salary Information file as per format to bank for payroll processing for there Employees.

**Salary Information File:**

Hr user will select Company, month, and year. After submitting the Salary Information will be created and got attached to Document. Users can download files and can send them to Bank for salary processing.

> It will also show a separate section for the Qatar region if the employer is not the payer

screencapture-frappe-erpnext-local-8000-app-salary-information-file-HR-SIF-2021-00003-2021-08-05-16_15_20 copy

#### WPS implementation for Qatar

**Company:**
As per SIF format, we also need to capture payer details too if the Employer is not Payer.

Screenshot 2021-07-30 at 4 00 29 PM

**Salary Structure:**
As per the new format of SIF, we also Need to capture Housing Allowance, Food Allowance, transportation allowance, etc,  

Screenshot 2021-07-30 at 3 55 59 PM

**CSV Format for Qatar:**
image

#### WPS implementation for UAE

**company:**
Users can define Employer establishment ID and Employer Reference Number as asked by the bank in SIF. 
image  
> Establishment id field is also present for Qatar Region.

**Employee:**
Agent Id field in salary details in Employee master. This is the routing code that is assigned to the Employee’s Bank/ Agent where their account is held. All UAE Banks and Exchanges have their routing codes, so you can obtain a list from banks

**Salary Component:**
As per the new format of SIF, We also need to Capture Fixed salary components and variable salary component, which is asked by the bank for salary Processing.

Screenshot 2021-08-04 at 3 22 43 PM

**CSV Format for UAE:**
image

**Some new Sections for capturing some details and Notifications related to that.**

**Some regional in company master fields for Qatar:**

image

image

image

For Family Visa:
image

For Individual Visa: 
image

**Notification:**
- On Passport Expiry
- On Driving Licenses Expiry

### CODE REVIEW

**Comment by bibinqcs** [comment]
I cant wait to test this. Awesome work @Anurag810 

Just from the picture you shared for the UAE sif format, few points for your reference ( 😅 just in case it slipped through)

> EDR,1235678909,123781263812,56710373873929,2021-07-01,2021-07-31,31.0,0,0,0.0
> SDR,21434665756,UIB,AT483200000012345864,21-08-04,1420,072021,1,AED,2938923920302

For the EDR record 
- 1235678909 should be the Ministry of Labour ID (MOL ID) I went through the proposed code changes I could not see a custom field for Ministry of Labour ID
- 31.0 the Days in period only needs the whole number
- ~~0.0  the leave without pay days only needs the whole number~~ **Update:** On all online resources it just mentions to mark '0' if there are no leave without pay days but 🤔 there is no mention of how it would handle Half Days which are leave without pay, I will confirm this and provide an update

For the Salary Control Record in UAE it is SCR not SDR
- UIB,AT483200000012345864 from what I understood it looks like the bank short name and IBAN number. it is not needed just like the EDR the company AGENT ID (routing code) is only needed
- 21-08-04 should be 2021-08-04
- after the total EDR count the total salary i.e Total Fixed and Total Variable is missing

should look like this

EDR,1235678909,123781263812,56710373873929,2021-07-01,2021-07-31,31,1000.00,500.00,0
EDR,1235678910,123781263812,56710373873930,2021-07-01,2021-07-31,31,1000.00,500.00,0
SCR,21434665756,987654321,2021-08-04,1420,072021,2,3000.00,AED,2938923920302

you can refer this sources

https://www.habibbank.com/uae/downloads/HBZwps_crg.pdf (Page 34 - Page 35)
https://www.dib.ae/docs/default-source/wps-guidelines/wps-file-format-reference-guide_v1-1.pdf (page 1-2)
https://www.emiratesislamic.ae/eng/assets/files/businessbankingforms/EI_businessONLINE_Salary_WPS_File_Creation_User_Guide.pdf (page 2-3)

File: erpnext/hr/doctype/employee/employee.json [hunk 1]
```diff
@@ -807,13 +815,117 @@
...
+   "collapsible": 1,
+   "fieldname": "visa_details_section",
+   "fieldtype": "Section Break",
+   "label": "Visa Details"
+  },
+  {
+   "fieldname": "visa_type",
+   "fieldtype": "Select",
+   "label": "Visa Type",
+   "options": "Family\nWork Permit\nVisit Visa\nTemporary Work Permit"
```
Comment (bibinqcs) [inline erpnext/hr/doctype/employee/employee.json:886]: Could you change _Work Permit_ to **Employment Visa** and _Temporary Work Permit_ to **Temporary Employment Visa**, in UAE the Ministry of Labour also issues an ID called Work Permit which is used in WPS sif report

**Comment by bibinqcs** [comment]
Could you add a notification for the expiry of Visa for all types, along with passport and Driving license this is a commonly used notification

**Comment by Anurag810** [comment]
@bibinqcs thanks for the review will work on the requested changes.

**Comment by mujeerhashmi** [comment]
Any updates, when is this going to be merged ?

**Comment by ankush** [comment]
Screenshot 2021-11-30 at 7 35 36 PM


@mohammedyusufshaikh this should really be a single commit. 😅 It will be painful in the future to look at history only to find 6 consecutive commits doing 1 thing. 

Use `git commit --amend` to amend last commit and/or `git rebase -i HEAD~N` to interactive squash last `N` commits. 

Also all linter fixes should ideally go in relevant commit or at very least clubbed in a single commit.

**Pull request closed without merging**